src/write_mongodb.c

```diff
@@ -55,179 +51,239 @@ struct wm_node_s {
+  BSON_APPEND_UTF8(ret, "type", vl->type);
+  BSON_APPEND_UTF8(ret, "type_instance", vl->type_instance);
 
-  bson_append_start_array(ret, "values"); /* {{{ */
+  BSON_APPEND_ARRAY_BEGIN(ret, "values", &subarray); /* {{{ */
   for (int i = 0; i < ds->ds_num; i++) {
     char key[16];
 
     ssnprintf(key, sizeof(key), "%i", i);
 
     if (ds->ds[i].type == DS_TYPE_GAUGE)
-      bson_append_double(ret, key, vl->values[i].gauge);
+      BSON_APPEND_DOUBLE(&subarray, key, vl->values[i].gauge);
     else if (store_rates)
-      bson_append_double(ret, key, (double)rates[i]);
+      BSON_APPEND_DOUBLE(&subarray, key, (double)rates[i]);
     else if (ds->ds[i].type == DS_TYPE_COUNTER)
-      bson_append_long(ret, key, vl->values[i].counter);
+      BSON_APPEND_INT64(&subarray, key, vl->values[i].counter);
     else if (ds->ds[i].type == DS_TYPE_DERIVE)
-      bson_append_long(ret, key, vl->values[i].derive);
+      BSON_APPEND_INT64(&subarray, key, vl->values[i].derive);
     else if (ds->ds[i].type == DS_TYPE_ABSOLUTE)
-      bson_append_long(ret, key, vl->values[i].absolute);
-    else
-      assert(23 == 42);
+      BSON_APPEND_INT64(&subarray, key, vl->values[i].absolute);
+    else {
+      ERROR("write_mongodb plugin: Unknown ds_type %d for index %d",
+            ds->ds[i].type, i);
+      bson_free(ret);
+      return NULL;
+    }
   }
-  bson_append_finish_array(ret); /* }}} values */
+  bson_append_array_end(ret, &subarray); /* }}} values */
 
-  bson_append_start_array(ret, "dstypes"); /* {{{ */
+  BSON_APPEND_ARRAY_BEGIN(ret, "dstypes", &subarray); /* {{{ */
   for (int i = 0; i < ds->ds_num; i++) {
     char key[16];
 
     ssnprintf(key, sizeof(key), "%i", i);
 
     if (store_rates)
-      bson_append_string(ret, key, "gauge");
+      BSON_APPEND_UTF8(&subarray, key, "gauge");
     else
-      bson_append_string(ret, key, DS_TYPE_TO_STRING(ds->ds[i].type));
+      BSON_APPEND_UTF8(&subarray, key, DS_TYPE_TO_STRING(ds->ds[i].type));
   }
-  bson_append_finish_array(ret); /* }}} dstypes */
+  bson_append_array_end(ret, &subarray); /* }}} dstypes */
 
-  bson_append_start_array(ret, "dsnames"); /* {{{ */
+  BSON_APPEND_ARRAY_BEGIN(ret, "dsnames", &subarray); /* {{{ */
   for (int i = 0; i < ds->ds_num; i++) {
     char key[16];
 
     ssnprintf(key, sizeof(key), "%i", i);
-    bson_append_string(ret, key, ds->ds[i].name);
+    BSON_APPEND_UTF8(&subarray, key, ds->ds[i].name);
   }
-  bson_append_finish_array(ret); /* }}} dsnames */
-
-  bson_finish(ret);
+  bson_append_array_end(ret, &subarray); /* }}} dsnames */
 
   sfree(rates);
-  return (ret);
+  return ret;
 } /* }}} bson *wm_create_bson */
 
-static int wm_write(const data_set_t *ds, /* {{{ */
-                    const value_list_t *vl, user_data_t *ud) {
-  wm_node_t *node = ud->data;
-  char collection_name[512];
-  bson *bson_record;
-  int status;
+static int wm_initialize(wm_node_t *node) /* {{{ */
+{
+  char *uri;
+  int uri_length;
+  char *format_string;
```

Comment:
some static code analyzers may freak out when a `char*` is used as a format string. can you either use a define or a `char const *`, please.